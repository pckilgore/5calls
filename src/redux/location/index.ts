export {
  LocationAction,
  LocationActionType,
  LocationClearedAction,
  LocationSetAction,
  SetLocationFetchTypeAction,
  SetSplitDistrictAction,
  CacheCityAction,
  NewLocationLookupAction
} from './action';
export {
  setLocation,
  clearAddress,
  setCachedCity,
  setLocationFetchType,
  setSplitDistrict
} from './actionCreator';
export { LocationState, locationStateReducer } from './reducer';
