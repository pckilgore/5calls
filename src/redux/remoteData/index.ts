export {
  IssuesAction,
  RemoteDataAction,
  RemoteDataActionType,
  CallCountAction
} from './action';
export { issuesActionCreator, callCountActionCreator } from './actionCreator';
export { RemoteDataState, remoteDataReducer } from './reducer';
export {
  startup,
  getIssuesIfNeeded,
  fetchCallCount
} from './asyncActionCreator';
