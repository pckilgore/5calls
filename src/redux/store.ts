import {
  applyMiddleware,
  createStore,
  Store,
  compose,
  Middleware
} from 'redux';
import { persistStore, Persistor } from 'redux-persist';
import rootReducer, { ApplicationState, DefaultApplicationState } from './root';
import thunk from 'redux-thunk';

const middlewares: Middleware[] = [thunk];

export let persistor = {} as Persistor;
export let store = {} as Store<ApplicationState>;

export default initialState => {
  store = createStore(
    rootReducer,
    DefaultApplicationState,
    compose(
      applyMiddleware(...middlewares),
      // This added for Redux Dev Tools - install Chrome or Firefox extension to use
      typeof window === 'object' &&
        // tslint:disable-next-line: no-string-literal
        typeof window['devToolsExtension'] !== 'undefined'
        ? // tslint:disable-next-line: no-string-literal
          window['devToolsExtension']()
        : f => f
    )
  ) as Store<ApplicationState>;

  persistor = persistStore(store);

  return store;
};
