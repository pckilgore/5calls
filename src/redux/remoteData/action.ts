import { RemoteDataAction } from './action';
import { Action } from 'redux';
import { Issue, ContactList } from '../../common/models';

export enum RemoteDataActionType {
  GET_ISSUES = 'GET_ISSUES',
  GET_CONTACTS = 'GET_CONTACTS',
  GET_CALL_TOTAL = 'GET_CALL_TOTAL'
}

export interface RemoteDataAction extends Action {
  type: RemoteDataActionType;
  payload?: {};
}

export interface IssuesAction extends RemoteDataAction {
  type: RemoteDataActionType.GET_ISSUES;
  payload: Issue[];
}

export interface ContactsAction extends RemoteDataAction {
  type: RemoteDataActionType.GET_CONTACTS;
  payload: ContactList;
}

export interface CallCountAction extends RemoteDataAction {
  type: RemoteDataActionType.GET_CALL_TOTAL;
  payload: number;
}
