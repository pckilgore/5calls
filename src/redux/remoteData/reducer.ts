import { Reducer } from 'redux';
import { Issue, ContactList, Contact } from '../../common/models';
import { RemoteDataAction, RemoteDataActionType } from './index';

export const defaultRemoteDataState: RemoteDataState = {
  issues: [],
  inactiveIssues: [],
  contacts: new ContactList(),
  callTotal: 0,
  errorMessage: ''
};

export interface RemoteDataState {
  issues: Issue[];
  inactiveIssues: Issue[];
  contacts: ContactList;
  callTotal: number;
  errorMessage: string;
}

export const remoteDataReducer: Reducer<RemoteDataState> = (
  state: RemoteDataState = {} as RemoteDataState,
  action: RemoteDataAction
): RemoteDataState => {
  switch (action.type) {
    case RemoteDataActionType.GET_ISSUES:
      // filter all issues into active / inactive
      let activeIssues: Issue[] = [];
      let inactiveIssues: Issue[] = [];
      if (action && action.payload) {
        let issues = action.payload as Issue[];
        activeIssues = issues
          .filter(item => {
            return item.active === true;
          })
          .map(item => Object.assign(new Issue(), item));
        inactiveIssues = issues
          .filter(item => {
            return item.active === false;
          })
          .map(item => Object.assign(new Issue(), item));
      }

      const issuesState = Object.assign({}, state, {
        issues: activeIssues,
        inactiveIssues: inactiveIssues
      });
      return issuesState;
    case RemoteDataActionType.GET_CONTACTS:
      const contactList = action.payload as ContactList;
      contactList.representatives = contactList.representatives.map(contact =>
        Object.assign(new Contact(), contact)
      );
      return Object.assign({}, state, {
        contacts: contactList
      });
    case RemoteDataActionType.GET_CALL_TOTAL:
      return Object.assign({}, state, { callTotal: action.payload });
    default:
      return state;
  }
};
