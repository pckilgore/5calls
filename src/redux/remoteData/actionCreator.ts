import { CallCountAction, IssuesAction, RemoteDataActionType } from './index';
import { Issue, ContactList } from '../../common/models';
import { ContactsAction } from './action';

export const issuesActionCreator = (issues: Issue[]): IssuesAction => {
  return {
    type: RemoteDataActionType.GET_ISSUES,
    payload: issues
  };
};

export const callCountActionCreator = (callTotal: number): CallCountAction => {
  return {
    type: RemoteDataActionType.GET_CALL_TOTAL,
    payload: callTotal
  };
};

export const contactsActionCreator = (
  contacts: ContactList
): ContactsAction => {
  return {
    payload: contacts,
    type: RemoteDataActionType.GET_CONTACTS
  };
};
