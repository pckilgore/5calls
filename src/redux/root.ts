import storage from 'redux-persist/lib/storage';
import getStoredState from 'redux-persist/lib/integration/getStoredStateMigrateV4';

import { ApplicationState } from './root';
import { LocationState, locationStateReducer } from './location';
import { CallState, callStateReducer } from './callState';
import { RemoteDataState, remoteDataReducer } from './remoteData';
import { UserStatsState, userStatsReducer } from './userStats';
import { UserState, userStateReducer } from './userState';
import { persistCombineReducers } from 'redux-persist';
import { defaultRemoteDataState } from './remoteData/reducer';
import { defaultUserStatsState } from './userStats/reducer';

export enum OutcomeType {
  UNAVAILABLE = 'unavailable',
  VOICEMAIL = 'voicemail',
  CONTACT = 'contact',
  SKIP = 'skip'
}

export interface ApplicationState {
  remoteDataState: RemoteDataState;
  callState: CallState;
  locationState: LocationState;
  userStatsState: UserStatsState;
  userState: UserState;
}

export const DefaultApplicationState: ApplicationState = {
  remoteDataState: defaultRemoteDataState,
  callState: {} as CallState,
  locationState: {} as LocationState,
  userStatsState: defaultUserStatsState,
  userState: {} as UserState
};

// DANGER: TypeScript magic ahead!!
// These articles should help your understanding of this section:
// https://blog.mariusschulz.com/2017/01/06/typescript-2-1-keyof-and-lookup-types
// https://blog.mariusschulz.com/2017/01/20/typescript-2-1-mapped-types
// https://moin.world/2017/06/18/10-TypeScript-Features-You-Might-Not-Know/
//
// This will be used by redux-persist in store.ts
// A type that is an ApplicationState key
export type ApplicationStateKeyType = keyof ApplicationState;
// A TS mapped type to type out an object shape that represents
// all ApplicationState keys and their string equivalents.
type ApplicationStateKeyTypes = {
  [K in ApplicationStateKeyType]: ApplicationStateKeyType
};
// Object that maps out all possible ApplicationState keys
export const ApplicationStateKey: ApplicationStateKeyTypes = {
  locationState: 'locationState',
  remoteDataState: 'remoteDataState',
  callState: 'callState',
  userStatsState: 'userStatsState',
  userState: 'userState'
};

const v4config = {
  whitelist: [
    ApplicationStateKey.locationState,
    ApplicationStateKey.userStatsState,
    ApplicationStateKey.userState,
    ApplicationStateKey.callState
  ]
};

const config = {
  key: 'fivecalls',
  debug: true,
  storage,
  whitelist: [
    ApplicationStateKey.locationState,
    ApplicationStateKey.userStatsState,
    ApplicationStateKey.userState,
    ApplicationStateKey.callState
  ],
  getStoredState: getStoredState(v4config)
};

const rootReducer = persistCombineReducers(config, {
  remoteDataState: remoteDataReducer,
  callState: callStateReducer,
  locationState: locationStateReducer,
  userStatsState: userStatsReducer,
  userState: userStateReducer
});

export default rootReducer;
