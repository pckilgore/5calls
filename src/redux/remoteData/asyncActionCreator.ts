import { Dispatch } from 'redux';
import * as querystring from 'querystring';

import { CountData, Issue, ContactList } from './../../common/models';
import {
  getAllIssues,
  getCountData,
  postBackfillOutcomes,
  getUserCallDetails,
  getContacts,
  noLocationError
} from '../../services/apiServices';
import { setCachedCity } from '../location/index';
import { issuesActionCreator, callCountActionCreator } from './index';
import { clearContactIndexes } from '../callState/';
import { ApplicationState } from '../root';
import { LoginService, UserProfile } from '@5calls/react-components';
import { Auth0Config } from '../../common/constants';
import { UserContactEvent } from '../userStats';
import { setUploadedActionCreator } from '../userStats/actionCreator';
import {
  clearProfileActionCreator,
  setAuthTokenActionCreator,
  setProfileActionCreator
} from '../userState';
import { setInvalidAddress, setLocation } from '../location/actionCreator';
import { store } from '../store';
import { contactsActionCreator } from './actionCreator';
import { removeURLParameter } from '../../components/shared/utils';
import bugsnagClient from '../../services/bugsnag';

export const getIssuesIfNeeded = () => {
  const state = store.getState();

  // Only make the api call if it hasn't already been made
  if (
    !state.remoteDataState.issues ||
    state.remoteDataState.issues.length === 0
  ) {
    getAllIssues()
      .then((response: Issue[]) => {
        store.dispatch(issuesActionCreator(response));
      })
      .catch(error => {
        // tslint:disable-next-line:no-console
        console.error(`error getting issues: ${error.message}`, error);
      });
  }
};

export const getContactsIfNeeded = (force: boolean) => {
  const state = store.getState();

  if (state.remoteDataState.contacts.representatives.length === 0 || force) {
    getContacts()
      .then((contactList: ContactList) => {
        store.dispatch(contactsActionCreator(contactList));
        store.dispatch(setCachedCity(contactList.location));
        store.dispatch(setInvalidAddress(false));
      })
      .catch(error => {
        if (error === noLocationError) {
          // normal case for when we don't yet have a location
        } else {
          // tslint:disable-next-line:no-console
          console.error('couldnt fetch contacts: ', error);
          bugsnagClient.notify(
            Error('reps error with location ' + state.locationState.address)
          );
          store.dispatch(setInvalidAddress(true));
        }
      });
  }
};

export const fetchCallCount = () => {
  return getCountData()
    .then((response: CountData) => {
      store.dispatch(callCountActionCreator(response.count));
    })
    .catch(error =>
      // tslint:disable-next-line:no-console
      console.error(`fetchCallCount error: ${error.message}`, error)
    );
};

export const fetchDonations = () => {
  return (
    dispatch: Dispatch<ApplicationState>,
    getState: () => ApplicationState
  ) => {
    return;
    // return getDonations()
    //   .then((response: DonationGoal) => {
    //     const donations: Donations = response.goal;
    //     dispatch(donationsActionCreator(donations));
    //   })
    //   // tslint:disable-next-line:no-console
    //   .catch(e => console.error(`fetchDonations error: ${e.message}`, e));
  };
};

export const uploadStatsIfNeeded = () => {
  return (
    dispatch: Dispatch<ApplicationState>,
    getState: () => ApplicationState
  ) => {
    const state: ApplicationState = getState();

    if (state.userState.idToken) {
      let unuploadedStats: UserContactEvent[] = [];

      for (let i = 0; i < state.userStatsState.all.length; i++) {
        if (!state.userStatsState.all[i].uploaded) {
          unuploadedStats.push(state.userStatsState.all[i]);
          dispatch(setUploadedActionCreator(state.userStatsState.all[i].time));
        }
      }

      if (unuploadedStats.length > 0) {
        postBackfillOutcomes(unuploadedStats, state.userState.idToken);
      }
    }
  };
};

export interface UserCallDetails {
  stats: UserStats;
  weeklyStreak: number;
  firstCallTime: number;
  calls: DailyCallReport[];
}

export interface DailyCallReport {
  date: string;
  issues: IssueSummary[];
}

export interface IssueSummary {
  count: number;
  issue_name: string;
}

export interface UserStats {
  voicemail: number;
  unavailable: number;
  contact: number;
}

export const getProfileInfo = async (): Promise<UserProfile> => {
  const state = store.getState();

  if (state.userState.profile && state.userState.idToken) {
    const callDetails = await getUserCallDetails(state.userState.idToken);
    // attach details to token profile
    let filledProfile = state.userState.profile;
    filledProfile.callDetails = callDetails;

    return filledProfile;
  } else {
    // not logged in
  }

  return Promise.reject('no profile sorry');
};

export const startup = () => {
  const state = store.getState();

  // clear contact indexes loaded from local storage
  store.dispatch(clearContactIndexes());

  // check expired login and handle or logout
  const auth = new LoginService(Auth0Config);
  if (state.userState.profile && state.userState.idToken) {
    auth
      .checkAndRenewSession(state.userState.profile, state.userState.idToken)
      .then(authResponse => {
        // Set the updated profile ourselves - auth is a component that doesn't know about redux
        store.dispatch(setAuthTokenActionCreator(authResponse.authToken));
        store.dispatch(setProfileActionCreator(authResponse.userProfile));
      })
      .catch(error => {
        // clear the session
        store.dispatch(clearProfileActionCreator());
      });
  }

  fetchCallCount();

  // set location automatically if it comes in via query string
  let replacedLocation = false;
  let q = querystring.parse(location.search.substr(1));
  const newLocation = q['setLocation']; // tslint:disable-line:no-string-literal
  if (newLocation !== undefined && newLocation.length > 0) {
    store.dispatch(setLocation(newLocation));
    replacedLocation = true;
  }
  // remove query string param so we don't set it again
  if (replacedLocation) {
    window.history.replaceState(
      {},
      document.title,
      removeURLParameter(location.href, 'setLocation')
    );
  }
};
