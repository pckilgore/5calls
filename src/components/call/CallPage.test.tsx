import * as React from 'react';
import { shallow } from 'enzyme';
import { History, Location } from 'history';
import i18n from '../../services/i18n';
import { I18nextProvider } from 'react-i18next';
import { CallPageWithRouter } from './index';
import { CallState } from '../../redux/callState';
import { RemoteDataState } from '../../redux/remoteData';

test('snapshot should render correctly with an issue and NO group', () => {
  const pageProps = initPage();
  const component = shallow(
    <I18nextProvider i18n={i18n}>
      <CallPageWithRouter {...pageProps} />
    </I18nextProvider>
  );
  expect(component).toMatchSnapshot();
});

const initPage = () => {
  return {
    match: { params: { issueid: '100' }, isExact: true, path: '', url: '' },
    location: {} as Location,
    history: {} as History,
    callState: {} as CallState,
    remoteState: {} as RemoteDataState
  };
};
