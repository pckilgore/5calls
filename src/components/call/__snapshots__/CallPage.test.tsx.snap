// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should render correctly with an issue and NO group 1`] = `
<withRouter(CallPageView)
  callState={Object {}}
  history={Object {}}
  location={Object {}}
  match={
    Object {
      "isExact": true,
      "params": Object {
        "issueid": "100",
      },
      "path": "",
      "url": "",
    }
  }
  remoteState={Object {}}
/>
`;
